lib/knetwork/dns.go

```diff
@@ -0,0 +1,191 @@
+	return func(s *DnsServer) error {
+		s.domains = domains
+		return nil
+	}
+}
+
+func WithListener(l net.Listener) DNSModifier {
+	return func(s *DnsServer) error {
+		s.Listener = l
+		return nil
+	}
+}
+
+func WithHost(ip string) DNSModifier {
+	return func(s *DnsServer) error {
+		s.host = ip
+		return nil
+	}
+}
+
+type DnsServer struct {
+	dnsServer *dns.Server
+	sync.RWMutex
+	routeMap map[string][]string
+	Port     int
+	Logger   *klog.Logger
+	domains  []string
+	Listener net.Listener
+	host     string
+}
+
+func (s *DnsServer) Start() error {
+	mux := dns.NewServeMux()
+	for _, domain := range s.domains {
+		mux.HandleFunc(dns.Fqdn(domain), s.HandleIncoming)
+	}
+	s.dnsServer = &dns.Server{Handler: mux}
+	if s.Listener != nil {
+		s.dnsServer.Listener = s.Listener
+		s.dnsServer.Net = "udp"
+		s.dnsServer.Addr = s.Listener.Addr().String()
+	}
+	if s.Port != 0 {
+		s.dnsServer.Addr = net.JoinHostPort(s.host, strconv.Itoa(s.Port))
+	}
+	return s.dnsServer.ListenAndServe()
+}
+
+func (s *DnsServer) Stop() error {
+	return s.dnsServer.Shutdown()
+}
+
+func (s *DnsServer) AddEntry(name string, ips []string) error {
+	if s.routeMap[dns.Fqdn(name)] != nil {
+		return fmt.Errorf("%w: %s", DNSEntryExistError, name)
+	}
+	s.Lock()
+	s.routeMap[dns.Fqdn(name)] = ips
+	s.Unlock()
+	return nil
+}
+
+// ReplaceEntry will hard replace an entry. Consider it a force AddEntry
+func (s *DnsServer) ReplaceEntry(name string, ips []string) error {
+	s.Lock()
+	s.routeMap[name] = ips
+	s.Unlock()
+	return nil
+}
+
+// AppendToEntry will add ips to an existing entry. Any collisions are automatically handled. It will not automatically
+// add the entry if it does not exist
+func (s *DnsServer) AppendToEntry(name string, ips []string) error {
+	if s.routeMap[dns.Fqdn(name)] == nil {
+		return fmt.Errorf("%w: %s", DNSEntryNotExistError, name)
+	}
+	fqdnName := dns.Fqdn(name)
+	s.Lock()
+	s.routeMap[fqdnName] = appendIfNotPresent(s.routeMap[fqdnName], ips)
+	s.Unlock()
+	return nil
+}
+
+// RemoveIPFromEntry will remove ips from an entry if the entry and name exists. I will not error if the ip to delete
+// is not found in th records
+func (s *DnsServer) RemoveIPFromEntry(name string, ips []string) error {
+	if s.routeMap[name] == nil {
+		return fmt.Errorf("%w: %s", DNSEntryNotExistError, name)
+	}
+	return nil
+}
+
+func (s *DnsServer) RemoveEntry(name string) error {
+	return nil
+}
+
+func (s *DnsServer) HandleIncoming(writer dns.ResponseWriter, incoming *dns.Msg) {
+	m := &dns.Msg{}
+	m.SetReply(incoming)
+	m.Compress = false
+	switch incoming.Opcode {
+	case dns.OpcodeQuery:
+		s.ParseDNS(m)
+	}
+	err := writer.WriteMsg(m)
+	if err != nil {
+		s.Logger.Errorw(err.Error())
+	}
+}
+
+func (s *DnsServer) ParseDNS(m *dns.Msg) {
+	for _, q := range m.Question {
+		switch q.Qtype {
+		case dns.TypeA:
+			ips := s.routeMap[q.Name]
+			for _, ip := range ips {
+				rr, err := dns.NewRR(fmt.Sprintf("%s A %s", q.Name, ip))
```

Comment:
don't parse the string every time you need to generate a response, parse the string when adding it to the map. Keep the rr in the map directly.

Rationale:
1) If an invalid record is added, you want to find out at addition time, not when the response is generated.
2) Parsing and sprintf is expensive, no reason to spend the cycles here.

Personally, I would change `routeMap map[string][]string` to instead be `routeMap map[string]*Record`, and then you can define record as:

```
type Record struct {
  IPs []dns.RR
  IPv6s []dns.RR
  TXTs []string
}
```
(IPs is the list of things returned for an 'A' Query, IPv6s for 'AAAA', TXTs for TXT query. TXT may come particularly handy to add annotations about the IPs returned - eg, description of machine, or records for scripts to use)